app/models/ont/run.rb

```diff
@@ -3,12 +3,69 @@
 module Ont
   # Ont::Run
   class Run < ApplicationRecord
-    include Stateful
+    include Uuidable
+
+    enum state: { pending: 0, completed: 1, user_terminated: 2, instrument_crashed: 3, restart: 4 }
+
+    NAME_PREFIX = 'ONTRUN-' # Used for generating a unique experiment name for the run.
+
+    # This association creates a link to the instrument this run belongs to.
+    # We are setting to avoid automatic detection of inverse association from the instrument.
+    # XXX: Do we need a default instrument to initialise the run with?
+    belongs_to :instrument,
+               class_name: 'Ont::Instrument',
+               foreign_key: :ont_instrument_id,
+               inverse_of: false
 
     has_many :flowcells, foreign_key: :ont_run_id, inverse_of: :run, dependent: :destroy
 
+    accepts_nested_attributes_for :flowcells, allow_destroy: true
+
     scope :active, -> { where(deactivated_at: nil) }
 
+    validate :check_max_number_of_flowcells, :check_flowcell_position, :check_flowcell_pool
+
+    def check_flowcell_position
+      return if instrument.blank?
+
+      position_set = Set.new
+      flowcells.each do |flowcell|
+        next if flowcell.position.blank?
+
+        if flowcell.position < 1 || flowcell.position > instrument.max_number_of_flowcells
+          errors.add(:flowcells, "position #{flowcell.position} is out of range for the instrument")
+        elsif position_set.include? flowcell.position
+          errors.add(:flowcells, "position #{flowcell.position} is duplicated in the same run")
+        else
+          position_set.add(flowcell.position)
+        end
+      end
+    end
+
+    def check_flowcell_pool
+      pool_set = Set.new
+      flowcells.each do |flowcell|
+        next if flowcell.ont_pool_id.blank?
+
+        if pool_set.include? flowcell.ont_pool_id
+          errors.add(:flowcells, "pool #{flowcell.ont_pool_id} is duplicated in the same run")
+        else
+          pool_set.add(flowcell.ont_pool_id)
+        end
+      end
+    end
+
+    def check_max_number_of_flowcells
+      return if instrument.blank?
```

Comment:
can merge these guard clauses into one something like return if instrument.blank? || flowcells.length <= intrument.max_number_of_flowcells

Reply:
also if you validate instrument's presence you wouldnt need the instrument.blanks guard

Reply:
Instrument presence is validated implicitly by belongs_to.
Without the instrument we cannot use those some validators because of instrument.max_number_of_flowcells (attribute access on nil). If ont_instrument_id is not given, the client would get 500 error about attribute access on nil instead of missing instrument id.

The guard is not part of the validation. It checks whether that validation can be done.